acados/ocp_nlp/ocp_nlp_constraints_bgp.c

```diff
@@ -0,0 +1,1334 @@
+                    &work->tmp_ni, nb);
+
+    // TODO(andrea): nz > 0 supported, but Hessian contribution associated with algebraic variables is neglected. 
+
+    // nonlinear
+    if (nphi > 0)
+    {
+        struct blasfeo_dvec_args x_in;  // input x of external fun;
+        x_in.x = memory->ux;
+        x_in.xi = nu;
+
+        struct blasfeo_dvec_args u_in;  // input u of external fun;
+        u_in.x = memory->ux;
+        u_in.xi = 0;
+
+        struct blasfeo_dvec_args z_in;  // input z of external fun;
+        z_in.x = memory->z_alg;
+        z_in.xi = 0;
+
+        ext_fun_type_in[0] = BLASFEO_DVEC_ARGS;
+        ext_fun_in[0] = &x_in;
+        ext_fun_type_in[1] = BLASFEO_DVEC_ARGS;
+        ext_fun_in[1] = &u_in;
+        ext_fun_type_in[2] = BLASFEO_DVEC_ARGS;
+        ext_fun_in[2] = &z_in;
+
+        struct blasfeo_dvec_args fun_out;
+        fun_out.x = &work->tmp_ni;
+        fun_out.xi = nb + ng;
+        ext_fun_type_out[0] = BLASFEO_DVEC_ARGS;
+        ext_fun_out[0] = &fun_out;  // fun: nphi
+
+        struct blasfeo_dmat_args jac_phi_tran_out;
+        jac_phi_tran_out.A = memory->DCt;
+        jac_phi_tran_out.ai = 0;
+        jac_phi_tran_out.aj = ng;
+        ext_fun_type_out[1] = BLASFEO_DMAT_ARGS;
+        ext_fun_out[1] = &jac_phi_tran_out;  // jac': (nu+nx) * nphi
+
+        struct blasfeo_dmat_args jac_phi_z_tran_out; // Jacobian dphidz treated separately
+
+        jac_phi_z_tran_out.A = &work->tmp_nz_nphi;
+        jac_phi_z_tran_out.ai = 0;
+        jac_phi_z_tran_out.aj = 0;
+        ext_fun_type_out[2] = BLASFEO_DMAT_ARGS;
+        ext_fun_out[2] = &jac_phi_z_tran_out;  // jac': nz * nphi
+
+        struct blasfeo_dmat_args hess_out;
+        hess_out.A = &work->tmp_nr_nphi_nr;
+        hess_out.ai = 0;
+        hess_out.aj = 0;
+        ext_fun_type_out[3] = BLASFEO_DMAT_ARGS;
+        ext_fun_out[3] = &hess_out;  // hess: nphi * nr * nr
+
+        model->nl_constr_phi_fun_jac->evaluate(model->nl_constr_phi_fun_jac, ext_fun_type_in, ext_fun_in, ext_fun_type_out, ext_fun_out);
+        
+        // expand phi:
+        // phi(x, u, z) ~
+        // phi(\bar{x}, \bar{u}, \bar{z}) + 
+        // dphidx*(x - \bar{x}) + 
+        // dphidu*(u - \bar{u}) + 
+        // dphidz*(z - \bar{z}) =
+        //
+        // phi(\bar{x}, \bar{u}, \bar{z}) - dphidz*dzdx*\bar{x} - dphidz*dzdu*\bar{u} + 
+        // (dphidx + dphidz*dzdx)*(x - \bar{x}) +  
+        // (dphidu + dphidz*dzdu)*(u - \bar{u})  
+        
+        // update DCt
+        blasfeo_dgemm_nn(nu+nx, nphi, nz, 1.0, memory->dzduxt, 0, 0, &work->tmp_nz_nphi, 0, 0, 0.0, &work->tmp_nv_nphi, 0, 0, &work->tmp_nv_nphi, 0, 0);
+        blasfeo_dgead(nu+nx, nphi, 1.0, &work->tmp_nv_nphi, 0, 0, memory->DCt, ng, 0);	
+        // update memory->fun	
+        blasfeo_dgemv_t(nu+nx, nphi, -1.0, &work->tmp_nv_nphi, 0, 0, memory->ux, 0, 1.0, &memory->fun, 0, &memory->fun, 0);
+    }
+
+    if (nr > 0)
```

Comment:
How can it be nphi>0 and nr=0?

Reply:
Just wait a moment before rushing with changes, I think it should be discussed a bit more the overall approach :p

Reply:
OK, but simply $nr$ cannot be equal to zero.